fdbrpc/tests/fdbrpc_bench.actor.cpp

```diff
@@ -0,0 +1,232 @@
+/*
+ * fdbrpc_bench.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2024 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "flow/flow.h"
+#include "flow/Platform.h"
+#include "flow/TLSConfig.actor.h"
+#include "fdbrpc/fdbrpc.h"
+#include "fdbrpc/FlowTransport.h"
+#include <iostream>
+
+namespace fdbrpc_bench {
+NetworkAddress serverAddress;
+
+enum FdbRpcBenchWellKnownEndpoints {
+	WLTOKEN_ECHO_SERVER = WLTOKEN_FIRST_AVAILABLE,
+	WLTOKEN_COUNT_ENDPOINTS,
+};
+
+struct EchoServerInterface {
+	constexpr static FileIdentifier file_identifier = 3152015;
+	RequestStream<struct GetInterfaceRequest> getInterface;
+	RequestStream<struct EchoRequest> echo;
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, echo);
+	}
+};
+
+struct GetInterfaceRequest {
+	constexpr static FileIdentifier file_identifier = 12004156;
+	ReplyPromise<EchoServerInterface> reply;
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, reply);
+	}
+};
+
+struct EchoRequest {
+	constexpr static FileIdentifier file_identifier = 10624019;
+	std::string message;
+	ReplyPromise<std::string> reply;
+
+	template <class Ar>
+	void serialize(Ar& ar) {
+		serializer(ar, message, reply);
+	}
+};
+
+class StatCounter {
+public:
+	StatCounter(int size = 10) : vals(size) {}
+
+	int avg() {
+		int now_ts = this->now() / 1000;
+		int sum = 0;
+		for (auto [ts, v] : vals) {
+			if (ts < now_ts - vals.size())
+				continue;
+			sum += v;
+		}
+		return sum / vals.size();
+	}
+
+	int64_t now() {
+		auto n = std::chrono::system_clock::now();
+		auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(n.time_since_epoch());
+		return duration.count();
+	}
+
+	void inc() {
+		int ts = this->now() / 1000;
+		int pos = ts % vals.size();
+
+		auto [old_ts, v] = vals[pos];
+		if (old_ts < ts) {
+			vals[pos] = { ts, 1 };
+		} else {
+			vals[pos] = { old_ts, v + 1 };
+		}
+	}
+
+private:
+	std::vector<std::pair<int64_t, int>> vals;
```

Comment:
what's stored here? I don't follow what this `StatCounter` does. It seems each entry in the vector keeps a count, but I don't understand why the avg() is dividing the sum by `vals.size()`.

Reply:
Each entry in vector is all the of of ops for that second. We keep last X seconds of data. So dividing that by vals.size() to get ops/seconds.